base/reduce.jl

```diff
@@ -277,6 +277,76 @@ end
 mapreduce_impl(f, op, A::AbstractArrayOrBroadcasted, ifirst::Integer, ilast::Integer) =
     mapreduce_impl(f, op, A, ifirst, ilast, pairwise_blocksize(f, op))
 
+# repeat expr n times in a block
+macro _repeat(n::Int, expr)
+    Expr(:block, fill(esc(expr), n)...)
+end
+
+# the following mapreduce_impl is called by mapreduce for non-array iterators,
+# and implements an index-free in-order pairwise strategy:
+function mapreduce_impl(f, op, ::_InitialValue, itr)
+    it = iterate(itr)
+    it === nothing && return mapreduce_empty_iter(f, op, itr)
+    a1, state = it
+    it = iterate(itr, state)
+    it === nothing && return mapreduce_first(f, op, a1)
+    a2, state = it
+    v = op(f(a1), f(a2))
+    # unroll a few iterations to reduce overhead for small iterators:
+    @_repeat 14 begin
+        it = iterate(itr, state)
+        it === nothing && return v
+        a, state = it
+        v = op(v, f(a))
+    end
+    n = max(2, pairwise_blocksize(f, op))
+    v, state = _mapreduce_impl(f, op, itr, v, state, n-16)
+    while state !== nothing
+        v, state = _mapreduce_impl(f, op, itr, v, state, n)
+        n *= 2
+    end
+    return v
+end
+
+# apply reduction to at most ≈ n elements of itr, in a pairwise recursive fashion,
+# returning op(nt, ...n elements...), state --- state=nothing if itr ended
+function _mapreduce_impl(f, op, itr, nt, state, n)
+    # for the pairwise algorithm we want to reduce this block
+    # separately *before* combining it with nt, so we peel
+    # off the first element to initialize the block reduction:
+    it = iterate(itr, state)
+    it === nothing && return nt, nothing
+    a1, state = it
+    it = iterate(itr, state)
+    it === nothing && return op(nt, f(a1)), nothing
+    a2, state = it
+    v = op(f(a1), f(a2))
+
+    if n ≤ max(2, pairwise_blocksize(f, op)) # coarsened base case
+        @simd for _ = 3:n
+            it = iterate(itr, state)
+            it === nothing && return op(nt, v), nothing
+            a, state = it
+            v = op(v, f(a))
+        end
+        return op(nt, v), state
+    else
+        n >>= 1
+        v, state = _mapreduce_impl(f, op, itr, v, state, n-2)
+        state === nothing && return op(nt, v), nothing
+        v, state = _mapreduce_impl(f, op, itr, v, state, n)
+        # break return statement into two cases to help type inference
+        return state === nothing ? (op(nt, v), nothing) : (op(nt, v), state)
+    end
+end
+
+# for an arbitrary initial value, we need to call foldl,
+# because op(nt, itr[i]) may have a different type than op(nt, itr[j]))
+# … it's not clear how to reliably do this without foldl associativity.
```

Comment:
The problem is that re-association will introduce the initial value multiple times. e.g. if I do
```
sum(1:10, init=1)
```
then the result depends on how many times I split the pairwise reduction. i.e. if that is just
```
foldl(+, 1:10, init=1)
```
you get `56`. But if you do a pairwise split in the middle you get
```
(foldl(+, 1:5, init=1) + foldl(+, 5:10, init=1))
```
then we have a result of `57`. This is similar to but distinct from the problems you encounter when `op` is not associative. In this case, it turns out to be important that `init` must be an identity element under `op`, i.e. you must have `op(init, x) ≈ x`

Reply:
> But if you do a pairwise split in the middle you get
> 
> ```
> (foldl(+, 1:5, init=1) + foldl(+, 5:10, init=1))
> ```

Dumb question: why couldn't we do `foldl(+, 1:5, init=1) + foldl(+, 5:10)`?  Isn't this already doing `foldl` for the first 14 operations anyway?

...

_Oh._ Right, I suppose this is where type comes into play.  Because it'd be very weird and problematic if `reduce(+, Int8.(1:100), init=big(0))` did half of its calculation mod $2^8$ and half in arbitrary precision.  Yeah, ok.

Reply:
> Because it'd be very weird and problematic if `reduce(+, Int8.(1:100), init=big(0))` did half of its calculation mod $2^8$ and half in arbitrary precision.

But on the other hand, isn't that what you sign up for with `reduce`?  Since we very explicitly do not guarantee an evaluation order, I don't think you should be able to rely upon this sort of "cascading-promotion-from-the-init-value".

Reply:
what I infer from the docstring is that `init` must be the identity element of the monoid defined by `op` and `eltype(itr)`

so I might expect it to "work" when `typeof(op(init, x)) <: eltype(itr)` for all `x in itr`

but in the example `reduce(+, Int8.(1:100), init=big(0))` since the promoted return type of `+(::BigInt, ::Int8)` does not subtype `Int8` , I do not think that can reasonably be expected to always produce a sane result

Reply:
Yeah the docstring seems to give us a lot of leeway to either drop the `init` or use it each time and give a 'surprising' result if it's not the identity element.

Reply:
Let's continue this conversation in #53871, e.g. see my comment at https://github.com/JuliaLang/julia/pull/53871#issuecomment-2023165390